subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/resolver/KotlinBuildScriptModelRepository.kt

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.kotlin.dsl.resolver
+
+import org.gradle.kotlin.dsl.concurrent.RessurectingThread
+import org.gradle.kotlin.dsl.tooling.models.KotlinBuildScriptModel
+
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.suspendCoroutine
+
+
+private
+typealias AsyncModelRequest = Pair<KotlinBuildScriptModelRequest, Continuation<KotlinBuildScriptModel?>>
+
+
+internal
+open class KotlinBuildScriptModelRepository {
+
+    open suspend fun scriptModelFor(request: KotlinBuildScriptModelRequest): KotlinBuildScriptModel? =
+        suspendCoroutine { k ->
+            accept(request, k)
+            processor.poke()
+        }
+
+    protected
+    open fun accept(request: KotlinBuildScriptModelRequest, k: Continuation<KotlinBuildScriptModel?>) {
+        q.push(request to k)
+    }
+
+    protected
+    open fun fetch(request: KotlinBuildScriptModelRequest): KotlinBuildScriptModel =
+        fetchKotlinBuildScriptModelFor(request)
+
+    private
+    val q = ConcurrentGroupingQueue<AsyncModelRequest> {
+        first.scriptFile == it.first.scriptFile && first.projectDir == it.first.projectDir
+    }
+
+    private
+    val processor = RessurectingThread("Kotlin DSL Resolver") {
+        while (true) {
```

Comment:
Do you want to do `while (!Thread.currentThread().isInterupted)` here?

Reply:
Thanks for the suggestion, It might provide slightly better behaviour, yes. Notice that `nextGroup` already waits on a lock condition so the thread should already be nicely interruptible.